File: java/programmers/src/이모티콘_할인행사.java
public class 이모티콘_할인행사 {
    class TmpPrice{
        int price;
        int percent;
        TmpPrice(int x,int y){
            this.price = x;
            this.percent = y;
        }
    }
    class Solution {
        int[] percent = {10, 20, 30, 40};
        int[] cloneEmoticons;
        TmpPrice[] tmpPrice;
        int[] answer = new int[2];

        public int[] solution(int[][] users, int[] emoticons) {
            answer[0] = 0;
            answer[1] = 0;
            tmpPrice = new TmpPrice[emoticons.length];
            cloneEmoticons = emoticons.clone();
            bf(0, users);

            return answer;
        }

        public void bf(int index, int[][] users) {
            if (index == cloneEmoticons.length) {
                checkUser(users);
                return;
            }
            for (int i = index; i < cloneEmoticons.length; i++) {
                for (int j = 0; j < 4; j++) {
                    tmpPrice[i] = new TmpPrice(cloneEmoticons[i] -
                            cloneEmoticons[i] * percent[j] / 100, (j + 1) * 10);
                    bf(i + 1, users);
                }
            }
        }

        public void checkUser(int[][] users) {
            int totalPrice = 0;
            int totalPlus = 0;
            for (int i = 0; i < users.length; i++) {
                int userPrice = 0;
                for (int j = 0; j < tmpPrice.length; j++) {
                    if (tmpPrice[j].percent >= users[i][0]) {
                        userPrice += tmpPrice[j].price;
                        if (userPrice >= users[i][1]) {
                            userPrice = 0;
                            totalPlus++;
                            break;
                        }
                    }
                }
                totalPrice += userPrice;
            }
            if (totalPlus > answer[0]) {
                answer[0] = totalPlus;
                answer[1] = totalPrice;
                return;
            }
            if (totalPlus == answer[0] && totalPrice > answer[1]) {
                answer[1] = totalPrice;
            }


        }
    }
}

File: java/baekjoon/src/_이모티콘할인행사.kt
class _이모티콘할인행사 {
    class Solution {
        val discountRate = intArrayOf(0, 10, 20, 30, 40)
        var emoticonCount = 0
        lateinit var globalEmoticons: IntArray
        val answer = IntArray(2)
        lateinit var gUsers: Array<IntArray>

        fun solution(
            users: Array<IntArray>,
            emoticons: IntArray,
        ): IntArray {
            gUsers = users
            emoticonCount = emoticons.size
            globalEmoticons = emoticons

            checkDiscountedEmoticons(mutableListOf<Emoticon>(), 0)
            return answer
        }

        fun checkDiscountedEmoticons(
            discountedEmoticons: MutableList<Emoticon>,
            index: Int,
        ) {
            if (index == emoticonCount) {
                checkEveryUser(discountedEmoticons)
                return
            }

            for (rate in discountRate) {
                val e = globalEmoticons.get(index)
                val discountedEmoticon = Emoticon(e - e * rate / 100, rate)

                discountedEmoticons.add(discountedEmoticon)
                checkDiscountedEmoticons(discountedEmoticons, index + 1)
                discountedEmoticons.removeLast()
            }
        }

        class Emoticon(
            val price: Int,
            val discountRate: Int,
        )

        // 유저마다 체크
        fun checkEveryUser(discountedE: MutableList<Emoticon>) {
            var totalPayed = 0L
            var emoticonPlusCount = 0
            for (user in gUsers) {
                val rateForBuying = user[0]
                val minPrice = user[1]

                var payed = 0L
                for (e in discountedE) {
                    if (e.discountRate >= rateForBuying) {
                        payed += e.price
                    }
                }

                if (payed >= minPrice) {
                    payed = 0
                    emoticonPlusCount++
                }
                totalPayed += payed
            }

            if (answer[0] < emoticonPlusCount) {
                answer[0] = emoticonPlusCount
                answer[1] = totalPayed.toInt()
                // println("new ${answer[0]} - ${answer[1]}")
                discountedE.forEach {
                    // println("${it.price} / ${it.discountRate}")
                }
            } else if (answer[0] == emoticonPlusCount) {
                answer[1] = max(answer[1], totalPayed.toInt())
                // println("new ${answer[0]} - ${answer[1]}")
                discountedE.forEach {
                    // println("${it.price} / ${it.discountRate}")
                }
            }
        }
    }
}
